test/emqx_shared_sub_SUITE.erl

```diff
@@ -314,6 +313,102 @@ t_uncovered_func(_) ->
     ignored = emqx_shared_sub ! ignored,
     {mnesia_table_event, []} = emqx_shared_sub ! {mnesia_table_event, []}.
 
+t_per_group_config(_) ->
+    ok = ensure_group_config(#{
+                               <<"local_group_fallback">> => local,
+                               <<"local_group">> => local,
+                               <<"round_robin_group">> => round_robin,
+                               <<"sticky_group">> => sticky
+                              }),
+    %% Each test is repeated 4 times because random strategy may technically pass the test
+    %% so we run 8 tests to make random pass in only 1/256 runs
+
+    test_two_messages(sticky, <<"sticky_group">>),
+    test_two_messages(sticky, <<"sticky_group">>),
+    test_two_messages(round_robin, <<"round_robin_group">>),
+    test_two_messages(round_robin, <<"round_robin_group">>),
+    test_two_messages(sticky, <<"sticky_group">>),
+    test_two_messages(sticky, <<"sticky_group">>),
+    test_two_messages(round_robin, <<"round_robin_group">>),
+    test_two_messages(round_robin, <<"round_robin_group">>).
+
+t_local(_) ->
+    Node = start_slave('local_shared_sub_test', 21884),
+    GroupConfig = #{
+                    <<"local_group_fallback">> => local,
+                    <<"local_group">> => local,
+                    <<"round_robin_group">> => round_robin,
+                    <<"sticky_group">> => sticky
+                   },
+    ok = ensure_group_config(Node, GroupConfig),
+    ok = ensure_group_config(GroupConfig),
+
+    Topic = <<"local_foo/bar">>,
+    ClientId1 = <<"ClientId1">>,
+    ClientId2 = <<"ClientId2">>,
+
+    {ok, ConnPid1} = emqtt:start_link([{clientid, ClientId1}, {port, 21884}]),
+    {ok, ConnPid2} = emqtt:start_link([{clientid, ClientId2}]),
+
+    {ok, _} = emqtt:connect(ConnPid1),
+    {ok, _} = emqtt:connect(ConnPid2),
+
+    emqtt:subscribe(ConnPid1, {<<"$share/local_group/local_foo/bar">>, 0}),
+    emqtt:subscribe(ConnPid2, {<<"$share/local_group/local_foo/bar">>, 0}),
+
+    ct:sleep(100),
+
+    ?assertEqual(local, emqx_shared_sub:strategy(<<"local_group">>)),
+    ?assertEqual(local, rpc:call(Node, emqx_shared_sub, strategy, [<<"local_group">>])),
+
+    Message1 = emqx_message:make(ClientId1, 0, Topic, <<"hello1">>),
+    Message2 = emqx_message:make(ClientId2, 0, Topic, <<"hello2">>),
+
+    emqx:publish(Message1),
+    {true, UsedSubPid1} = last_message(<<"hello1">>, [ConnPid1, ConnPid2]),
+
+    rpc:call(Node, emqx, publish, [Message2]),
+    {true, UsedSubPid2} = last_message(<<"hello2">>, [ConnPid1, ConnPid2]),
+
+    ?assertNotEqual(UsedSubPid1, UsedSubPid2),
+
+    emqtt:stop(ConnPid1),
+    emqtt:stop(ConnPid2),
+    stop_slave(Node),
```

Comment:
It's safer to do these clean-ups in `end_per_testcase` because, if one of the assertions above fail, they won't run.

Reply:
I moved stops before the assertions